utils/s2n_mem.c

```diff
@@ -70,69 +93,35 @@ int s2n_realloc(struct s2n_blob *b, uint32_t size)
     /* blob already has space for the request */
     if (size < b->allocated) {
         b->size = size;
-        return 0;
+        return S2N_SUCCESS;
     }
 
-    void *data;
-    if (!use_mlock) {
-        data = realloc(b->data, size);
-        S2N_ERROR_IF(!data, S2N_ERR_ALLOC);
-
-        b->data = data;
-        b->size = size;
-        b->allocated = size;
-        b->growable = 1;
-        b->mlocked = 0;
-        return 0;
+    struct s2n_blob new_memory = {0};
+    if (s2n_get_memory(&new_memory, size) < 0) {
+        GUARD(s2n_free(&new_memory));
+        S2N_ERROR_PRESERVE_ERRNO();
     }
 
-    /* Page aligned allocation required for mlock */
-    uint32_t allocate = page_size * (((size - 1) / page_size) + 1);
-    S2N_ERROR_IF(posix_memalign(&data, page_size, allocate), S2N_ERR_ALLOC);
-
     if (b->size) {
-        memcpy_check(data, b->data, b->size);
+        memcpy_check(new_memory.data, b->data, b->size);
         GUARD(s2n_free(b));
     }
 
-    b->data = data;
-    b->size = size;
-    b->allocated = allocate;
-    b->growable = 1;
-
-#ifdef MADV_DONTDUMP
-    if (madvise(b->data, size, MADV_DONTDUMP) < 0) {
-        GUARD(s2n_free(b));
-        S2N_ERROR(S2N_ERR_MADVISE);
-    }
-#endif
-
-    if (mlock(b->data, size) < 0) {
-        GUARD(s2n_free(b));
-        S2N_ERROR(S2N_ERR_MLOCK);
-    }
-    b->mlocked = 1;
-
-    return 0;
+    *b = new_memory;
+    return S2N_SUCCESS;
 }
 
 int s2n_free(struct s2n_blob *b)
 {
     S2N_ERROR_IF(!s2n_blob_is_growable(b), S2N_ERR_FREE_STATIC_BLOB);
-    int munlock_rc = 0;
-    if (b->mlocked) {
-        munlock_rc = munlock(b->data, b->size);
-    }
-
+    /* To avoid memory leaks, still free the data even if we can't unlock / wipe it */
+    int zero_rc = s2n_blob_zero(b);
+    int munlock_rc = b->mlocked ? munlock(b->data, b->size) : 0;
```

Comment:
Munlock_rc is never read

Reply:
Its read here: https://github.com/awslabs/s2n/blob/7d3bdcf3618ad981127bbad613183644d41100e9/utils/s2n_mem.c#L122